include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h

```diff
@@ -512,31 +469,29 @@ struct __accessor
 };
 
 template <typename _ExecutionPolicy, typename _T>
-struct __storage
+class __storage_impl
 {
   private:
     using __sycl_buffer_t = sycl::buffer<_T, 1>;
-    ::std::shared_ptr<__sycl_buffer_t> __sycl_buf;
-    ::std::shared_ptr<_T> __usm_buf;
     bool __usm;
+    ::std::shared_ptr<_T> __usm_buf;
+    ::std::shared_ptr<__sycl_buffer_t> __sycl_buf;
 
   public:
-    __storage(_ExecutionPolicy& __exec, bool __u, ::std::size_t __n) : __usm(__u)
+    __storage_impl(_ExecutionPolicy& __exec, bool __u, ::std::size_t __n)
+        : __usm(__u),
+          __usm_buf(__u ? std::shared_ptr<_T>(
+                              __internal::__sycl_usm_alloc<_ExecutionPolicy, _T, sycl::usm::alloc::host>{__exec}(__n),
+                              __internal::__sycl_usm_free<_ExecutionPolicy, _T>{__exec})
+                        : decltype(__usm_buf){}),
+          __sycl_buf(__u ? decltype(__sycl_buf){} : ::std::make_shared<__sycl_buffer_t>(__sycl_buffer_t(__n)))
     {
-        if (__usm)
-        {
-            __usm_buf = std::shared_ptr<_T>(
-                __internal::__sycl_usm_alloc<_ExecutionPolicy, _T, sycl::usm::alloc::host>{__exec}(__n),
-                __internal::__sycl_usm_free<_ExecutionPolicy, _T>{__exec});
-        }
-        else
-            __sycl_buf = ::std::make_shared<__sycl_buffer_t>(__sycl_buffer_t(__n));
     }
```

Comment:
This code looks ugly and overcomplicated even comparing with the current implementation. What is the reason to refactor it like this? Essentially, that only optimizes away a default ctor and a move assignment for a shared pointer - is it a big win, really?

Reply:
The goal: better to initialize some smart pointers through constructor - I believe it's should work faster then assignment.

Reply:
This class looks like some common used functional inside `oneDPL` library so I think better to implement it as more effective.